src/gt4py/next/program_processors/runners/dace/workflow/translation.py

```diff
@@ -113,69 +113,125 @@ def _has_gpu_schedule(sdfg: dace.SDFG) -> bool:
 
-    Todo: Revisit this function once DaCe changes its behaviour in this regard.
+    The execution time is measured in seconds and represented as a 'float64' value.
+    It is returned from the SDFG as a one-element array in the '__return' data node.
     """
+    output, _ = sdfg.add_array("__return", [1], dace.float64)
+    start_time, _ = sdfg.add_scalar("gt_start_time", dace.float64, transient=True)
+    metrics_level = sdfg.add_symbol(gtx_wfdcommon.SDFG_ARG_METRIC_LEVEL, dace.int32)
+
+    #### 1. Set CUDA stram to default and create synchronization code for GPU target.
+    # Even when the target device is GPU, it can happen that dace emits code without
+    # GPU kernels. In this case, the cuda headers are not imported and the SDFG is
+    # compiled as plain C++. Therefore, we also check here the schedule of SDFG maps.
+    if gpu and _has_gpu_schedule(sdfg):
+        # If we are using the default stream, things are a bit simpler/harder. For some
+        #  reasons when using the default stream, DaCe seems to skip _all_ synchronization,
+        #  for more see [DaCe issue#2120](https://github.com/spcl/dace/issues/2120).
+        #  Thus the `CompiledSDFG.fast_call()` call is truly asynchronous, i.e. just
+        #  launches the kernels and then exits. Thus we have to add a synchronization
+        #  at the end to have a synchronous call. We cannot use `SDFG.append_exit_code()`
+        #  because that code is only run at the `exit()` stage, not after a call. Thus we
+        #  will generate an SDFGState that contains a Tasklet with the sync call.
+        assert dace.Config.get("compiler.cuda.max_concurrent_streams") == -1, (
+            f"Expected `max_concurrent_streams == -1` but it was `{dace.Config.get('compiler.cuda.max_concurrent_streams')}`."
+        )
 
-    if not gpu:
-        # This is only a problem on GPU. Dace uses OpenMP on CPU and
-        # the OpenMP parallel region creates a synchronization point.
-        return
-    elif not _has_gpu_schedule(sdfg):
-        # Even when the target device is GPU, it can happen that dace
-        # emits code without GPU kernels. In this case, the cuda headers
-        # are not imported and the SDFG is compiled as plain C++.
-        return
+        dace_gpu_backend = dace.Config.get("compiler.cuda.backend")
+        assert dace_gpu_backend in ["cuda", "hip"], f"GPU backend '{dace_gpu_backend}' is unknown."
 
-    assert dace.Config.get("compiler.cuda.max_concurrent_streams") == -1, (
-        f"Expected `max_concurrent_streams == -1` but it was `{dace.Config.get('compiler.cuda.max_concurrent_streams')}`."
-    )
+        # NOTE: We should actually wrap the `StreamSynchronize` function inside a
+        #   `DACE_GPU_CHECK()` macro. However, this only works in GPU context, but
+        #   here we are in CPU context. Thus we cannot do it.
+        sync_code = f"{dace_gpu_backend}StreamSynchronize({dace_gpu_backend}StreamDefault);"
```

Comment:
For a single stream is there a difference between `cudaStreamSynchronize` and `cudaDeviceSynchronize`? If not, we can replace this by `cudaDeviceSynchronize` and remove the checks that we have a single stream.